spec/lib/compiler_spec.rb

```diff
@@ -308,20 +308,106 @@
+
+        expect_compiled(%q{
+          def foo
+            %x{return bar(baz)}
+          end
+        }).to include("  return bar(baz);\n")
+
+        expect_compiled(%q{
+          def foo
+            %x{return bar(baz);}
+          end
+        }).to include("  return bar(baz);\n")
+
+        expect_compiled(%q{
+          def foo
+            %x{
+              return bar(baz)
+            }
+          end
+        }).to include("  return bar(baz);\n")
+
+        expect_compiled(%q{
+          def foo
+            %x{
+              return bar(baz);
+            }
+          end
+        }).to include("  return bar(baz);\n")
+
+        expect_compiled(%q{
+          def foo
+            789
+            `#{123 + bar} * 456;`
+          end
+        }).to include("  return $rb_plus(123, self.$bar()) * 456;\n")
+
+        expect_compiled(%q{
+          def foo
+            789
+            `#$bar * 456`
+          end
+        }).to include("  return $gvars.bar * 456;\n")
+
+        expect_compiled(%q{
+          def foo
+            789
+            `#@bar * 456`
+          end
+        }).to include("  return self.bar * 456;\n")
+
+        expect_compiled(%q{
+          if `compare === nil`
+            raise ArgumentError, "comparison of #{a.class} with #{b.class} failed"
+          end
+        }).to include("  if ($truthy(compare === nil)) {\n")
+
+        expect_compiled(%q{
+          def <<(count)
+            count = Opal.coerce_to! count, Integer, :to_int
+            `#{count} > 0 ? self << #{count} : self >> -#{count}`
+          end
+        }).to include("  return count > 0 ? self << count : self >> -count;\n")
+      end
+
+      it 'warns if a semicolon is used in a single line' do
+        expect_compiler(%{a = `1;`})              { |c| c.to receive(:warning).once }
+        expect_compiler(%{a = `1;`; return})      { |c| c.to receive(:warning).once }
+        expect_compiler(%{a = `1;`})              { |c| c.to receive(:warning).once }
+        expect_compiler(%{a = ` 1; `})            { |c| c.to receive(:warning).once }
+        expect_compiler(%{a = %x{\n 1; \n}})      { |c| c.to receive(:warning).once }
+        expect_compiler(%{def foo; ` 1;  `; end}) { |c| c.to receive(:warning).once }
+      end
+
+      it 'does not warn for statements' do
+        expect_compiler(%{`foo;`; return})       { |c| c.not_to receive(:warning) }
+        expect_compiler(%{`foo;`; 123})          { |c| c.not_to receive(:warning) }
+      end
+
+      it 'does not warn for multiline x-strings' do
+        expect_compiler(%{a = `1;\n2;`})          { |c| c.not_to receive(:warning) }
+        expect_compiler(%{a = `1;\n2;3; `})       { |c| c.not_to receive(:warning) }
+        expect_compiler(%{def foo;` 1;\n  `;end}) { |c| c.to receive(:warning).once }
+      end
+    end
 
-      compiler = Opal::Compiler.new('`1;`')
-      expect(compiler).not_to receive(:warning)
+    def expect_compiler(code)
```

Comment:
That's a very, very complex construction.

I like how it's written in the parser gem. There's a field on the parser called `diagnostics_consumer` (or something similar, can't remember) that gets called on every error or warning. In the "test" mode it simply collects all arguments (adapter pattern) and puts them to some static array (like `ScratchPad` in rubyspec). Later in the spec any matcher can directly use this array.

Reply:
I see, basically it would be like 

```rb
warnings = []
allow_any_instance_of(Opal::Compiler).to receive(:warning) { |*args| warnings << args }
expect(warnings.size).to eq(1)
# etc.
```

---

The reason I used this odd construct is that it worked well with backtrace being the expectation at the same exact line that generated it. Let me see if I can come up with anything better